server/lib/chat/chat-service.ts

```diff
@@ -589,20 +633,36 @@ export default class ChatService {
             userId: msg.userId,
             time: Number(msg.sent),
           })
-          users.push({ id: msg.userId, name: msg.userName })
+          userIds.add(msg.userId)
           break
 
         default:
           return assertUnreachable(msg.data)
       }
     }
 
-    const mentions = await findUsersById(Array.from(mentionIds))
+    const [users, userMentions, channelMentions] = await Promise.all([
+      findUsersById(Array.from(userIds)),
+      findUsersById(Array.from(userMentionIds)),
+      getChannelInfos(Array.from(channelMentionIds)),
+    ])
+
+    const deletedChannels =
+      channelMentionIds.size === channelMentions.length
+        ? []
+        : Array.from(
+            subtract(
+              channelMentionIds,
+              channelMentions.map(c => c.id),
+            ),
+          )
 
     return {
       messages,
       users,
-      mentions,
+      mentions: userMentions,
+      channelMentions: channelMentions.map(c => toChannelSummary(c).channelInfo),
```

Comment:
Given how much this seems to be how you use this over and over again, I really wonder if it wouldn't be better to write 3 separate methods and call them as you need them